appserver/payara-appserver-modules/payara-micro-cdi/src/main/java/fish/payara/micro/cdi/ClusteredCDIEventBus.java

```diff
@@ -0,0 +1,203 @@
+ * file that accompanied this code.
+ *
+ * Modifications:
+ * If applicable, add the following below the License Header, with the fields
+ * enclosed by brackets [] replaced by your own identifying information:
+ * "Portions Copyright [year] [name of copyright owner]"
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package fish.payara.micro.cdi;
+
+import fish.payara.appserver.micro.services.CDIEventListener;
+import fish.payara.appserver.micro.services.PayaraClusteredCDIEvent;
+import fish.payara.appserver.micro.services.PayaraInstance;
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.annotation.Annotation;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.annotation.Resource;
+import javax.enterprise.concurrent.ManagedExecutorService;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Event;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.EventMetadata;
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Inject;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import org.glassfish.api.invocation.ComponentInvocation;
+import org.glassfish.api.invocation.InvocationManager;
+import org.glassfish.internal.api.Globals;
+
+/**
+ *
+ * @author steve
+ */
+@ApplicationScoped
+public class ClusteredCDIEventBus implements CDIEventListener {
+
+    @Inject
+    private PayaraInstance runtime;
+
+    @Inject
+    private BeanManager bm;
+
+    @Resource
+    private ManagedExecutorService managedExecutorService;
+
+    private ComponentInvocation capturedInvocation;
+
+    private InvocationManager im;
+
+    private ClassLoader capturedClassLoader;
+    
+    private final static String INSTANCE_PROPERTY = "InstanceName";
+    
+    private final static String EVENT_PROPERTY = "EventName";
+    
+    @PostConstruct
+    void postConstruct() {
+        runtime.addCDIListener(this);
+        capturedClassLoader = Thread.currentThread().getContextClassLoader();
+        im = Globals.getDefaultHabitat().getService(InvocationManager.class);
+        capturedInvocation = im.getCurrentInvocation();
```

Comment:
Can't we use the Java EE ContextService to avoid using the hack with "captured" invocation manager? What is the issue here - is it because you are trying to access the executor from a hazelcast Java SE thread? I think that ContextService should enrich the current thread to enable using the executor.